src/sb_pointcloud_processing/src/IGVCVisualizerNode.cpp

```diff
@@ -0,0 +1,152 @@
+//
+// Created by robyn on 08/03/18.
+//
+
+#include "IGVCVisualizerNode.h"
+
+
+// Allocation of memory for static member variables
+float IGVCVisualizerNode::h_margin_of_error;
+float IGVCVisualizerNode::s_margin_of_error;
+float IGVCVisualizerNode::v_margin_of_error;
+
+boost::shared_ptr<pcl::visualization::PCLVisualizer> IGVCVisualizerNode::viewer;
+pcl::PointCloud<pcl::PointXYZRGB>::Ptr IGVCVisualizerNode::visualized_cloud;
+bool IGVCVisualizerNode::isPaused;
+
+IGVCVisualizerNode::IGVCVisualizerNode(int argc, char **argv, std::string node_name) {
+    // ROS setup
+    ros::init(argc, argv, node_name);
+    ros::NodeHandle nh;
+    ros::NodeHandle private_nh("~");
+
+    // Always start resumed
+    isPaused = false;
+
+    // Initialise visualiser settings
+    viewer = setUpPCLVisualizer();
+
+    // Initialise margin of error and image update rate
+    retrieveVisualizerParameters(private_nh);
+
+    // Setup subscriber
+    std::string pcl_topic = "/zed/point_cloud/cloud_registered";
+    uint32_t queue_size = 1;
+    image_sub = nh.subscribe<sensor_msgs::PointCloud2>(pcl_topic, queue_size, &IGVCVisualizerNode::rawPCLCallBack,
+                                                       this);
+
+    // Setup timer callback for updating the camera
+    timer = nh.createTimer(ros::Duration(image_update_rate), &IGVCVisualizerNode::updateVisualizerCallback, this);
+}
+
+void IGVCVisualizerNode::rawPCLCallBack(const sensor_msgs::PointCloud2::ConstPtr &input) {
+    // Only update point cloud if system is not paused
+    if (!isPaused) {
+        // Obtain the ROS pointcloud and convert into PCL Pointcloud2
+        pcl::PCLPointCloud2::Ptr pcl_input(new pcl::PCLPointCloud2);
+        pcl_conversions::toPCL(*(input), *(pcl_input));
+
+        // Converts from the Pointcloud2 format to PointcloudRGB
+        pcl::PointCloud<pcl::PointXYZRGB>::Ptr pcl_rgb(
+                new pcl::PointCloud<pcl::PointXYZRGB>());
+        pcl::fromPCLPointCloud2(*pcl_input, *pcl_rgb);
+
+        // Remove NaNs from point cloud
+        std::vector<int> indices;
+        pcl::removeNaNFromPointCloud(*pcl_rgb, *pcl_rgb, indices);
+
+        // Renew pcl point data
+        visualized_cloud = pcl_rgb;
+
+        // Remove all old point clouds from the view
+        viewer->removeAllPointClouds(0);
+
+        // Show the new point cloud retrieved from the camera
+        viewer->addPointCloud(visualized_cloud, "visualized_cloud");
+
+        // Setup how big the points are in the point cloud
+        viewer->setPointCloudRenderingProperties(pcl::visualization::PCL_VISUALIZER_POINT_SIZE, 1.0,
+                                                 "visualized_cloud");
+    }
+}
+
+void IGVCVisualizerNode::updateVisualizerCallback(const ros::TimerEvent &event) {
+    // Update the viewer
+    viewer->spinOnce(100);
+    boost::this_thread::sleep(boost::posix_time::microseconds(100000));
```

Comment:
A quick comment here explaining why we need to use `boost::this_thread::sleep` would be helpful.

Reply:
!Resolved